### PR TITLE
Bugfix for approle delete from safe and Service account Managed by api

### PR DESCRIPTION
**Pull request by sarathmohan4**

### CODE REVIEW

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 1]
```diff
@@ -1694,4 +1692,173 @@ private void updateGroupPolicyAssociationOnSvcaccDelete(String svcAccName,Map<St
...
+     * @param userDetails
+     * @param owner
+     * @return
+     */
+    public ResponseEntity<String> getServiceAccountManagerDetails(String token, UserDetails userDetails, String owner) {
+        AndFilter andFilter = new AndFilter();
+		andFilter.and(new EqualsFilter("cn", owner));
+        andFilter.and(new EqualsFilter("objectClass", "user"));
+        List<String> managedServiceAccounts = getServiceAccountManagerDetails(andFilter);
+        return ResponseEntity.status(HttpStatus.OK).body("{ \"keys\": "+ managedServiceAccounts.get(0)+"}");
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Why do you call this as Keys?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 2]
```diff
@@ -1694,4 +1692,173 @@ private void updateGroupPolicyAssociationOnSvcaccDelete(String svcAccName,Map<St
...
+        log.debug(JSONUtil.getJSON(ImmutableMap.<String, String>builder().
+                put(LogMessage.USER, ThreadLocalContext.getCurrentMap().get(LogMessage.USER).toString()).
+                put(LogMessage.ACTION, "getServiceAccountManagerDetails").
+                put(LogMessage.MESSAGE, String.format("Trying to get manager details")).
+                put(LogMessage.APIURL, ThreadLocalContext.getCurrentMap().get(LogMessage.APIURL).toString()).
+                build()));
+        return userLdapTemplate.search("", filter.encode(), new AttributesMapper<String>() {
+            @Override
+            public String mapFromAttributes(Attributes attr) throws NamingException {
+				Map<String, String> managerDetails = new HashMap<>();
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Why Map? Can't this be a model object?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 3]
```diff
@@ -1694,4 +1692,173 @@ private void updateGroupPolicyAssociationOnSvcaccDelete(String svcAccName,Map<St
...
+	 * @param userDetails
+	 * @param token
+	 * @param serviceAccountApprole
+	 * @return
+	 */
+	public ResponseEntity<String> removeApproleFromSvcAcc(UserDetails userDetails, String token, ServiceAccountApprole serviceAccountApprole) {
+		log.debug(JSONUtil.getJSON(ImmutableMap.<String, String>builder().
+				put(LogMessage.USER, ThreadLocalContext.getCurrentMap().get(LogMessage.USER).toString()).
+				put(LogMessage.ACTION, "Delete Approle from Service Account").
+				put(LogMessage.MESSAGE, String.format ("Trying to delete approle from Service Account [%s]", serviceAccountApprole.getApprolename())).
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: remove or delete?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/SafesService.java [hunk 4]
```diff
@@ -1994,7 +1994,7 @@ private void deletePolicies (String token, Safe safe) {
 	public ResponseEntity<String> removeApproleFromSafe(String token, String jsonstr) {
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/SafesService.java]: removeApproleFromSafe - Is it for service account or safe?

Comment (sivanag1974) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/SafesService.java]: If it is for service account, is the existing method getting changed? If yes, what will happen to the existing implementation?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/v2/controller/ServiceAccountsControllerV2.java [hunk 5]
```diff
@@ -155,4 +171,18 @@
...
+    /**
+     * Get Manager details for the service account
+     * @param request
+     * @param token
+     * @param owner
+     * @return
+     */
+    @ApiOperation(value = "${ServiceAccountsControllerV2.getServiceAccountsMeta.value}", notes = "${ServiceAccountsControllerV2.getServiceAccountsMeta.notes}")
+    @GetMapping(value="/v2/ad/serviceaccounts/{owner}", produces="application/json")
+    public ResponseEntity<String> getServiceAccountManagerDetails(HttpServletRequest request, @RequestHeader(value="vault-token") String token, @PathVariable("owner" ) String owner){
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/v2/controller/ServiceAccountsControllerV2.java]: What is the use of this?

File: tvaultapi/src/main/resources/application.properties [hunk 6]
```diff
@@ -19,6 +19,7 @@ ad.userdn:
 ad.username:
 ad.password:
 ad.svc.acc.suffix=
+ad.baseusr:
```
Comment (sivanag1974) [inline tvaultapi/src/main/resources/application.properties]: Why baseusr?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/SafesService.java (hunk 4 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/SafesService.java]: This is removeApproleFromSafe. This is the fix for the bug: policy not getting deleted.

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/common/TVaultConstants.java [hunk 7]
```diff
@@ -83,4 +83,6 @@
     public static final String SVC_ACC_ROLES_METADATA_MOUNT_PATH = "metadata/ad/roles";
     public static final String SVC_ACC_ROLES_PATH = "ad/roles";
     public static final Long PASSWORD_AUTOROTATE_TTL_MAX_VALUE = 1590897977L;
+    public static final String SVC_ACC_NAME_PREFIX="svc";
```
[the remark this answers is not in the document]
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/common/TVaultConstants.java]: removed

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/common/TVaultConstants.java [hunk 8]
```diff
@@ -83,4 +83,6 @@
...
+    public static final String SVC_AD_USERDN = "OU=Production,OU=Users,OU=Accounts,DC=clouddev,DC=corporate,DC=t-mobile,DC=com";
```
[the remark this answers is not in the document]
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/common/TVaultConstants.java]: removed

File: tvaultapi/src/main/resources/application.properties (hunk 6 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/resources/application.properties]: Added for the base for fetching managed_by details

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/v2/controller/ServiceAccountsControllerV2.java (hunk 5 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/v2/controller/ServiceAccountsControllerV2.java]: refactored

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 3 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: remove, corrected

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 1 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: refactored

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 2 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: refactored

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/model/ADUserAccount.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+	 */
+	public void setUserName(String userName) {
+		this.userName = userName;
+	}
+	/* (non-Javadoc)
+	 * @see java.lang.Object#toString()
+	 */
+	@Override
+	public String toString() {
+		return "DirectoryUser [userId=" + userId + ", userEmail=" + userEmail + ", displayName=" + displayName
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/model/ADUserAccount.java]: Why do you still call it as DirectoryUser?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 10]
```diff
@@ -68,17 +65,25 @@
...
 	private static Logger log = LogManager.getLogger(ServiceAccountsService.class);
+	private final static String[] permissions = {"read", "write", "deny", "sudo"};
 
 	@Autowired
 	@Qualifier(value = "svcAccLdapTemplate")
 	private LdapTemplate ldapTemplate;
-	
+
+	@Autowired
+	@Qualifier(value = "userLdapTemplate")
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: userLdapTemplate - user means which user? ADUser or some other user?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 11]
```diff
@@ -124,6 +129,32 @@
 			}
 		}
 		List<ADServiceAccount> allServiceAccounts = getADServiceAccounts(andFilter);
+
+		// get the managed_by details
+		if (allServiceAccounts != null && !allServiceAccounts.isEmpty()) {
+			List<String> ownerlist = allServiceAccounts.stream().map(m -> m.getManagedBy().getUserName()).collect(Collectors.toList());
+			// remove duplicate usernames
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: What will happen if getUserName() is null?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 12]
```diff
@@ -124,6 +129,32 @@
...
+				filterQuery.append("(cn=" + owner + ")");
+			}
+			filterQuery.append("))");
+			List<ADUserAccount> managedServiceAccounts = getServiceAccountManagerDetails(filterQuery.toString());
+
+			// Update the managedBy withe ADUserAccount object
+			for (ADServiceAccount adServiceAccount : allServiceAccounts) {
+				if (!StringUtils.isEmpty(adServiceAccount.getManagedBy().getUserName())) {
+					List<ADUserAccount> adUserAccount = managedServiceAccounts.stream().filter(f -> f.getUserName().equalsIgnoreCase(adServiceAccount.getManagedBy().getUserName())).collect(Collectors.toList());
+					if (!adUserAccount.isEmpty()) {
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Will there be a possibility of NPE for f->getUserName()?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 13]
```diff
@@ -183,9 +214,11 @@ public ADServiceAccount mapFromAttributes(Attributes attr) throws NamingExceptio
 						adServiceAccount.setWhenCreated(instant);
 					}
 					if (attr.get("manager") != null) {
+						ADUserAccount adUserAccount = new ADUserAccount();
 						String managedByStr = (String) attr.get("manager").get();
 						String managedBy= managedByStr.substring(3, managedByStr.indexOf(","));
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Possibility of NPE and OutofBounds Exception in above two lines?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 14]
```diff
@@ -661,7 +683,7 @@ private String populateSvcAccMetaJson(String svcAccName, String username) {
 		String svcAccName = serviceAccountUser.getSvcAccName();
 		String access = serviceAccountUser.getAccess();
 
-		if(!ControllerUtil.areSvcUserInputsValid(serviceAccountUser)) {
+		if(!isSvcaccPermissionInputValid(serviceAccountUser.getAccess())) {
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: What is serviceAccountUser is null and will there be NPE?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 15]
```diff
@@ -680,7 +702,7 @@ private String populateSvcAccMetaJson(String svcAccName, String username) {
 		}
 		boolean isAuthorized = true;
 		if (userDetails != null) {
-			isAuthorized = canAddOrRemoveUser(userDetails, serviceAccountUser, TVaultConstants.ADD_USER);
+			isAuthorized = hasAddOrRemovePermission(userDetails, serviceAccountUser.getSvcAccName(), token);
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Again NPE for serviceAccountUser.getSvcAccName()?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 16]
```diff
@@ -1694,4 +1733,172 @@ private void updateGroupPolicyAssociationOnSvcaccDelete(String svcAccName,Map<St
...
+				put(LogMessage.ACTION, "Remove Approle from Service Account").
+				put(LogMessage.MESSAGE, String.format ("Trying to remove approle from Service Account [%s]", serviceAccountApprole.getApprolename())).
+				put(LogMessage.APIURL, ThreadLocalContext.getCurrentMap().get(LogMessage.APIURL).toString()).
+				build()));
+		String approleName = serviceAccountApprole.getApprolename();
+		String svcAccName = serviceAccountApprole.getSvcAccName();
+		String access = serviceAccountApprole.getAccess();
+
+		if (serviceAccountApprole.getApprolename().equals(TVaultConstants.SELF_SERVICE_APPROLE_NAME)) {
+			return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("{\"errors\":[\"Access denied: no permission to associate/remove this AppRole to any Service Account\"]}");
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: When you are trying to removeAppRoleFromSvcAcc, why do we need to show a generic message (associate/remove this AppRole to any Service Account)?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 17]
```diff
@@ -1694,4 +1733,172 @@ private void updateGroupPolicyAssociationOnSvcaccDelete(String svcAccName,Map<St
...
+				Map<String,String> params = new HashMap<String,String>();
+				params.put("type", "app-roles");
+				params.put("name",approleName);
+				params.put("path",path);
+				params.put("access","delete");
+				Response metadataResponse = ControllerUtil.updateMetadata(params,token);
+				if(metadataResponse !=null && (HttpStatus.NO_CONTENT.equals(metadataResponse.getHttpstatus()) || HttpStatus.OK.equals(metadataResponse.getHttpstatus()))){
+					log.debug(JSONUtil.getJSON(ImmutableMap.<String, String>builder().
+							put(LogMessage.USER, ThreadLocalContext.getCurrentMap().get(LogMessage.USER).toString()).
+							put(LogMessage.ACTION, "Remove AppRole to Service Account").
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Grammar?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 18]
```diff
@@ -1694,4 +1733,172 @@ private void updateGroupPolicyAssociationOnSvcaccDelete(String svcAccName,Map<St
...
+							put(LogMessage.MESSAGE, "AppRole configuration Success.").
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Incomplete log message

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 19]
```diff
@@ -1694,4 +1733,172 @@ private void updateGroupPolicyAssociationOnSvcaccDelete(String svcAccName,Map<St
...
+				if(approleControllerResp.getHttpstatus().equals(HttpStatus.NO_CONTENT)){
+					log.error(JSONUtil.getJSON(ImmutableMap.<String, String>builder().
+							put(LogMessage.USER, ThreadLocalContext.getCurrentMap().get(LogMessage.USER).toString()).
+							put(LogMessage.ACTION, "Remove AppRole from Service Account").
+							put(LogMessage.MESSAGE, "Reverting, approle policy update success").
+							put(LogMessage.RESPONSE, (null!=metadataResponse)?metadataResponse.getResponse():TVaultConstants.EMPTY).
+							put(LogMessage.STATUS, (null!=metadataResponse)?metadataResponse.getHttpstatus().toString():TVaultConstants.EMPTY).
+							put(LogMessage.APIURL, ThreadLocalContext.getCurrentMap().get(LogMessage.APIURL).toString()).
+							build()));
+					return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("{\"errors\":[\"Approle configuration failed.Please try again\"]}");
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Punctuation

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 20]
```diff
@@ -1694,4 +1733,172 @@ private void updateGroupPolicyAssociationOnSvcaccDelete(String svcAccName,Map<St
...
+				}else{
+					log.error(JSONUtil.getJSON(ImmutableMap.<String, String>builder().
+							put(LogMessage.USER, ThreadLocalContext.getCurrentMap().get(LogMessage.USER).toString()).
+							put(LogMessage.ACTION, "Remove AppRole from Service Account").
+							put(LogMessage.MESSAGE, "Reverting approle policy update failed").
+							put(LogMessage.RESPONSE, (null!=metadataResponse)?metadataResponse.getResponse():TVaultConstants.EMPTY).
+							put(LogMessage.STATUS, (null!=metadataResponse)?metadataResponse.getHttpstatus().toString():TVaultConstants.EMPTY).
+							put(LogMessage.APIURL, ThreadLocalContext.getCurrentMap().get(LogMessage.APIURL).toString()).
+							build()));
+					return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("{\"errors\":[\"Approle configuration failed.Contact Admin \"]}");
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Punctuation (space between "." and "Contact")

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java [hunk 21]
```diff
@@ -1694,4 +1733,172 @@ private void updateGroupPolicyAssociationOnSvcaccDelete(String svcAccName,Map<St
...
+				}
+			}
+			else {
+				return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("{\"errors\":[\"Failed to remove approle from the Service Account\"]}");
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Why INTERNAL_SERVER_ERROR?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/v2/controller/ServiceAccountsControllerV2.java [hunk 22]
```diff
@@ -137,11 +139,25 @@
...
     }
 
+	/**
+	 * Remove approle from Service Account
+	 * @param request
+	 * @param token
+	 * @param serviceAccountApprole
+	 * @return
+	 */
+	@ApiOperation(value = "${ServiceAccountsControllerV2.associateApprole.value}", notes = "${ServiceAccountsControllerV2.associateApprole.notes}")
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/v2/controller/ServiceAccountsControllerV2.java]: "${ServiceAccountsControllerV2.associateApprole.value}" -  Why do you still call it as associateAppRole when you are trying to removeApproleFromSvcAcc?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 14 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: serviceAccountUser cannot be null as there is api validation for the fields

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 15 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: serviceAccountUser will not be null as there is api validation for the fields

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 16 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: corrected

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 17 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: corrected

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 18 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: corrected

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 19 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Corrected

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 20 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Corrected

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/v2/controller/ServiceAccountsControllerV2.java (hunk 22 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/v2/controller/ServiceAccountsControllerV2.java]: Corrected

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 21 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Changed to Bad request

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/model/ADUserAccount.java (hunk 9 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/model/ADUserAccount.java]: Corrected

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 10 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Changed to adUserLdapTemplate

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 11 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Added removal of empty manager names from list

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 12 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: Added null check in filter

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java (hunk 13 above)
Comment (sarathmohan4) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/ServiceAccountsService.java]: NPE will not come here as the null check for manager attribute is done already. Added string empty check for substring

**Pull request merged**